tests/test_lora_fsdp.py

```diff
@@ -0,0 +1,190 @@
+        ensure_lora_initialized(model)
+
+        inputs = tokenizer("The future of AI is", return_tensors="pt")
+
+        # Output with adapters enabled
+        with torch.no_grad():
+            output_with_adapters = model(**inputs).logits.clone()
+
+        # Output with adapters disabled
+        with model.disable_adapter():
+            with torch.no_grad():
+                output_without_adapters = model(**inputs).logits.clone()
+
+        max_diff = torch.abs(output_with_adapters - output_without_adapters).max().item()
+        assert max_diff > 1e-3, f"Adapters not affecting output. Max diff: {max_diff}"
+
+        # Compare disabled output to base model (no PEFT)
+        base_model = AutoModelForCausalLM.from_pretrained("gpt2")
+        with torch.no_grad():
+            base_output = base_model(**inputs).logits
+
+        base_diff = torch.abs(output_without_adapters - base_output).max().item()
+        assert base_diff < 1e-5, f"Disabled adapters don't match base model. Diff: {base_diff}"
+
+    def test_disable_adapter_with_fsdp(self, setup_dist, tokenizer):
+        """
+        Verify that disable_adapter works under FSDP:
+          * does not raise the old requires_grad RuntimeError,
+          * actually changes the output when adapters are toggled.
+        """
+        if not torch.cuda.is_available():
+            pytest.skip("CUDA required for this test")
+
+        torch.cuda.manual_seed_all(42)
+
+        model = AutoModelForCausalLM.from_pretrained("gpt2")
+        config = LoraConfig(
+            task_type=TaskType.CAUSAL_LM,
+            r=16,
+            lora_alpha=32,
+            target_modules=["c_attn", "c_proj"],
+            init_lora_weights="gaussian",
+        )
+        model = get_peft_model(model, config)
+        ensure_lora_initialized(model)
+
+        # Wrap with FSDP
+        model = FSDP(model, use_orig_params=True).cuda()
+        inputs = {k: v.cuda() for k, v in tokenizer("The future of AI is", return_tensors="pt").items()}
+
+        # Ensure disable_adapter context does not produce the old requires_grad error
+        try:
+            with model.disable_adapter():
+                pass
+        except RuntimeError as e:
+            if "requires_grad" in str(e):
+                pytest.fail(f"Got requires_grad RuntimeError that should be fixed: {e}")
+            else:
+                raise
+
+        # Check that adapters actually change output
+        with torch.no_grad():
+            output_with_adapters = model(**inputs).logits.clone()
+
+        with model.disable_adapter():
+            with torch.no_grad():
+                output_without_adapters = model(**inputs).logits.clone()
+
+        max_diff = torch.abs(output_with_adapters - output_without_adapters).max().item()
+        assert max_diff > 1e-3, f"FSDP: Adapters not affecting output when disabled. Max diff: {max_diff}."
+
+    def test_modules_to_save_wrapper_handling(self, setup_dist, tokenizer):
```

Comment:
For completeness, let's add a copy of this test, but instead of checking `ModulesToSaveWrapper`, it tests `TrainableTokensWrapper` (from `peft.utils.other`). To enable it, pass `trainable_tokens=[1, 3, 5]` to `LoraConfig`.